File: python/numpy_version/simulation.py
import numpy as np
import constants as cs
from state import SimulationState
import poisson
import collisions

def step1_compute_electron_density(sim: SimulationState):
    x  = sim.x_e[:sim.N_e]
    c0 = x * cs.INV_DX
    p  = np.clip(c0.astype(np.int32), 0, cs.N_G - 2)
    p_f = p.astype(np.float64)
    w_left  = p_f + 1.0 - c0    # weight for left node
    w_right = c0 - p_f          # weight for right node

    sim.e_density[:] = 0.0
    np.add.at(sim.e_density, p,     w_left  * cs.FACTOR_W)
    np.add.at(sim.e_density, p + 1, w_right * cs.FACTOR_W)

    sim.e_density[0]         *= 2.0
    sim.e_density[cs.N_G - 1] *= 2.0
    sim.cumul_e_density       += sim.e_density


def step1_compute_ion_density(sim: SimulationState, t: int):
    if (t % cs.N_SUB) != 0:
        sim.cumul_i_density += sim.i_density
        return

    x  = sim.x_i[:sim.N_i]
    c0 = x * cs.INV_DX
    p  = np.clip(c0.astype(np.int32), 0, cs.N_G - 2)
    p_f = p.astype(np.float64)
    w_left  = p_f + 1.0 - c0
    w_right = c0 - p_f

    sim.i_density[:] = 0.0
    np.add.at(sim.i_density, p,     w_left  * cs.FACTOR_W)
    np.add.at(sim.i_density, p + 1, w_right * cs.FACTOR_W)

    sim.i_density[0]         *= 2.0
    sim.i_density[cs.N_G - 1] *= 2.0
    sim.cumul_i_density       += sim.i_density


def step2_solve_poisson(sim: SimulationState):
    rho = cs.E_CHARGE * (sim.i_density - sim.e_density)
    poisson.solve_poisson(sim, rho, sim.Time)


def step3_move_electrons(sim: SimulationState, t_index: int):
    x  = sim.x_e[:sim.N_e]
    vx = sim.vx_e[:sim.N_e]
    vy = sim.vy_e[:sim.N_e]
    vz = sim.vz_e[:sim.N_e]

    c0 = x * cs.INV_DX
    p  = np.clip(c0.astype(np.int32), 0, cs.N_G - 2)
    p_f = p.astype(np.float64)
    c1 = p_f + 1.0 - c0
    c2 = c0 - p_f
    e_x = c1 * sim.efield[p] + c2 * sim.efield[p + 1]

    if sim.measurement_mode:
        mean_v = vx - 0.5 * e_x * cs.FACTOR_E
        np.add.at(sim.counter_e_xt[:, t_index], p,     c1)
        np.add.at(sim.counter_e_xt[:, t_index], p + 1, c2)
        np.add.at(sim.ue_xt[:, t_index], p,     c1 * mean_v)
        np.add.at(sim.ue_xt[:, t_index], p + 1, c2 * mean_v)

        v_sqr  = mean_v**2 + vy**2 + vz**2
        energy = 0.5 * cs.E_MASS * v_sqr / cs.EV_TO_J
        np.add.at(sim.meanee_xt[:, t_index], p,     c1 * energy)
        np.add.at(sim.meanee_xt[:, t_index], p + 1, c2 * energy)

        e_idx  = np.minimum((energy / cs.DE_CS + 0.5).astype(np.int32), cs.CS_RANGES - 1)
        rate   = sim.sigma[cs.E_ION, e_idx] * np.sqrt(v_sqr) * cs.DT_E * cs.GAS_DENSITY
        np.add.at(sim.ioniz_rate_xt[:, t_index], p,     c1 * rate)
        np.add.at(sim.ioniz_rate_xt[:, t_index], p + 1, c2 * rate)

        # EEPF in center
        center_mask = (x > cs.MIN_X) & (x < cs.MAX_X)
        e_center = energy[center_mask]
        eepf_idx = (e_center / cs.DE_EEPF).astype(np.int32)
        valid = eepf_idx < cs.N_EEPF
        np.add.at(sim.eepf, eepf_idx[valid], 1.0)
        sim.mean_energy_accu_center    += float(np.sum(e_center))
        sim.mean_energy_counter_center += int(np.sum(center_mask))

    # Update velocity and position (in-place on views = modifies sim arrays)
    vx -= e_x * cs.FACTOR_E
    x  += vx  * cs.DT_E


def step4_move_ions(sim: SimulationState, t_index: int, t: int):
    if (t % cs.N_SUB) != 0:
        return

    x  = sim.x_i[:sim.N_i]
    vx = sim.vx_i[:sim.N_i]
    vy = sim.vy_i[:sim.N_i]
    vz = sim.vz_i[:sim.N_i]

    c0 = x * cs.INV_DX
    p  = np.clip(c0.astype(np.int32), 0, cs.N_G - 2)
    p_f = p.astype(np.float64)
    c1 = p_f + 1.0 - c0
    c2 = c0 - p_f
    e_x = c1 * sim.efield[p] + c2 * sim.efield[p + 1]

    if sim.measurement_mode:
        mean_v = vx + 0.5 * e_x * cs.FACTOR_I
        np.add.at(sim.counter_i_xt[:, t_index], p,     c1)
        np.add.at(sim.counter_i_xt[:, t_index], p + 1, c2)
        np.add.at(sim.ui_xt[:, t_index], p,     c1 * mean_v)
        np.add.at(sim.ui_xt[:, t_index], p + 1, c2 * mean_v)

        v_sqr  = mean_v**2 + vy**2 + vz**2
        energy = 0.5 * cs.AR_MASS * v_sqr / cs.EV_TO_J
        np.add.at(sim.meanei_xt[:, t_index], p,     c1 * energy)
        np.add.at(sim.meanei_xt[:, t_index], p + 1, c2 * energy)

    vx += e_x * cs.FACTOR_I
    x  += vx  * cs.DT_I


def step5_check_boundaries_electrons(sim: SimulationState):
    x = sim.x_e[:sim.N_e]

    mask_pow = x < 0.0
    mask_gnd = x > cs.L
    mask_out = mask_pow | mask_gnd

    sim.N_e_abs_pow += int(np.sum(mask_pow))
    sim.N_e_abs_gnd += int(np.sum(mask_gnd))

    mask_keep = ~mask_out
    n_keep = int(np.sum(mask_keep))

    sim.x_e[:n_keep]  = sim.x_e[:sim.N_e][mask_keep]
    sim.vx_e[:n_keep] = sim.vx_e[:sim.N_e][mask_keep]
    sim.vy_e[:n_keep] = sim.vy_e[:sim.N_e][mask_keep]
    sim.vz_e[:n_keep] = sim.vz_e[:sim.N_e][mask_keep]
    sim.N_e = n_keep


def step6_check_boundaries_ions(sim: SimulationState, t: int):
    if (t % cs.N_SUB) != 0:
        return

    x  = sim.x_i[:sim.N_i]
    vx = sim.vx_i[:sim.N_i]
    vy = sim.vy_i[:sim.N_i]
    vz = sim.vz_i[:sim.N_i]

    mask_pow = x < 0.0
    mask_gnd = x > cs.L
    mask_out = mask_pow | mask_gnd

    sim.N_i_abs_pow += int(np.sum(mask_pow))
    sim.N_i_abs_gnd += int(np.sum(mask_gnd))

    # Collect IFED for absorbed ions
    v_sqr  = vx**2 + vy**2 + vz**2
    energy = 0.5 * cs.AR_MASS * v_sqr / cs.EV_TO_J

    for mask, ifed in [(mask_pow, sim.ifed_pow), (mask_gnd, sim.ifed_gnd)]:
        e_abs = energy[mask]
        idx   = (e_abs / cs.DE_IFED).astype(np.int64)
        valid = idx < cs.N_IFED
        np.add.at(ifed, idx[valid], 1)

    mask_keep = ~mask_out
    n_keep = int(np.sum(mask_keep))

    sim.x_i[:n_keep]  = sim.x_i[:sim.N_i][mask_keep]
    sim.vx_i[:n_keep] = sim.vx_i[:sim.N_i][mask_keep]
    sim.vy_i[:n_keep] = sim.vy_i[:sim.N_i][mask_keep]
    sim.vz_i[:n_keep] = sim.vz_i[:sim.N_i][mask_keep]
    sim.N_i = n_keep


def step7_collisions_electrons(sim: SimulationState):
    if cs.USE_NULL_COLLISION:
        if sim.N_e == 0:
            return
        N_coll_star = int(sim.rng.binomial(sim.N_e, sim.P_star_e))
        if N_coll_star > sim.N_e:
            N_coll_star = sim.N_e
        if N_coll_star == 0:
            return

        candidates = sim.rng.choice(sim.N_e, size=N_coll_star, replace=False)

        vx = sim.vx_e[candidates]
        vy = sim.vy_e[candidates]
        vz = sim.vz_e[candidates]

        v_sqr = vx**2 + vy**2 + vz**2
        velocity = np.sqrt(v_sqr)
        energy = 0.5 * cs.E_MASS * v_sqr / cs.EV_TO_J
        e_idx = np.minimum((energy / cs.DE_CS + 0.5).astype(np.int32), cs.CS_RANGES - 1)

        real_nu = sim.sigma_tot_e[e_idx] * velocity
        p_accept = real_nu / sim.nu_star_e
        p_accept = np.minimum(p_accept, 1.0)

        rands = sim.rng.random(N_coll_star)
        accepted = rands < p_accept

        for idx, is_accepted in enumerate(accepted):
            if is_accepted:
                k = candidates[idx]
                collisions.collision_electron(sim, int(k), int(e_idx[idx]))
                sim.N_e_coll += 1
    else:
        vx = sim.vx_e[:sim.N_e]
        vy = sim.vy_e[:sim.N_e]
        vz = sim.vz_e[:sim.N_e]

        v_sqr    = vx**2 + vy**2 + vz**2
        velocity = np.sqrt(v_sqr)
        energy   = 0.5 * cs.E_MASS * v_sqr / cs.EV_TO_J
        e_idx    = np.minimum((energy / cs.DE_CS + 0.5).astype(np.int32), cs.CS_RANGES - 1)

        nu     = sim.sigma_tot_e[e_idx] * velocity
        p_coll = 1.0 - np.exp(-nu * cs.DT_E)

        rands     = sim.rng.random(sim.N_e)
        colliding = np.where(rands < p_coll)[0]

        for k in colliding:
            collisions.collision_electron(sim, int(k), int(e_idx[k]))
            sim.N_e_coll += 1


def step8_collisions_ions(sim: SimulationState, t: int):
    if (t % cs.N_SUB) != 0:
        return

    if cs.USE_NULL_COLLISION:
        if sim.N_i == 0:
            return
        N_coll_star = int(sim.rng.binomial(sim.N_i, sim.P_star_i))
        if N_coll_star > sim.N_i:
            N_coll_star = sim.N_i
        if N_coll_star == 0:
            return

        candidates = sim.rng.choice(sim.N_i, size=N_coll_star, replace=False)

        vx_a = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION, N_coll_star)
        vy_a = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION, N_coll_star)
        vz_a = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION, N_coll_star)

        vx = sim.vx_i[candidates]
        vy = sim.vy_i[candidates]
        vz = sim.vz_i[candidates]

        gx = vx - vx_a
        gy = vy - vy_a
        gz = vz - vz_a
        g_sqr = gx**2 + gy**2 + gz**2
        g = np.sqrt(g_sqr)

        energy = 0.5 * cs.MU_ARAR * g_sqr / cs.EV_TO_J
        e_idx = np.minimum((energy / cs.DE_CS + 0.5).astype(np.int32), cs.CS_RANGES - 1)

        real_nu = sim.sigma_tot_i[e_idx] * g
        p_accept = real_nu / sim.nu_star_i
        p_accept = np.minimum(p_accept, 1.0)

        rands = sim.rng.random(N_coll_star)
        accepted = rands < p_accept

        for idx, is_accepted in enumerate(accepted):
            if is_accepted:
                k = candidates[idx]
                collisions.collision_ion(
                    sim, int(k),
                    float(vx_a[idx]), float(vy_a[idx]), float(vz_a[idx]),
                    int(e_idx[idx])
                )
                sim.N_i_coll += 1
    else:
        vx = sim.vx_i[:sim.N_i]
        vy = sim.vy_i[:sim.N_i]
        vz = sim.vz_i[:sim.N_i]

        # Sample gas atom velocities for all ions at once
        vx_a = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION, sim.N_i)
        vy_a = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION, sim.N_i)
        vz_a = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION, sim.N_i)

        gx    = vx - vx_a
        gy    = vy - vy_a
        gz    = vz - vz_a
        g_sqr = gx**2 + gy**2 + gz**2
        g     = np.sqrt(g_sqr)

        energy = 0.5 * cs.MU_ARAR * g_sqr / cs.EV_TO_J
        e_idx  = np.minimum((energy / cs.DE_CS + 0.5).astype(np.int32), cs.CS_RANGES - 1)

        nu     = sim.sigma_tot_i[e_idx] * g
        p_coll = 1.0 - np.exp(-nu * cs.DT_I)

        rands     = sim.rng.random(sim.N_i)
        colliding = np.where(rands < p_coll)[0]

        for k in colliding:
            collisions.collision_ion(
                sim, int(k),
                float(vx_a[k]), float(vy_a[k]), float(vz_a[k]),
                int(e_idx[k])
            )
            sim.N_i_coll += 1


def step9_collect_xt_data(sim: SimulationState, t_index: int):
    if not sim.measurement_mode:
        return
    sim.pot_xt[:, t_index]    += sim.pot
    sim.efield_xt[:, t_index] += sim.efield
    sim.ne_xt[:, t_index]     += sim.e_density
    sim.ni_xt[:, t_index]     += sim.i_density


def do_one_cycle(sim: SimulationState, datafile_path: str = "conv.dat"):
    for t in range(cs.N_T):
        sim.Time += cs.DT_E
        t_index = t // cs.N_BIN

        step1_compute_electron_density(sim)
        step1_compute_ion_density(sim, t)
        step2_solve_poisson(sim)
        step3_move_electrons(sim, t_index)
        step4_move_ions(sim, t_index, t)
        step5_check_boundaries_electrons(sim)
        step6_check_boundaries_ions(sim, t)
        step7_collisions_electrons(sim)
        step8_collisions_ions(sim, t)
        step9_collect_xt_data(sim, t_index)

        if (t % 1000) == 0:
            print(f" c = {sim.cycle:8d}  t = {t:8d}  #e = {sim.N_e:8d}  #i = {sim.N_i:8d}")

    with open(datafile_path, "a") as f:
        f.write(f"{sim.cycle:8d}  {sim.N_e:8d}  {sim.N_i:8d}\n")

File: Go/2.algorithmic-port/simulation.go
package gopic

import (
	"fmt"
	"math"
)

//----------------------------------------------------------------------//
// initialization of the simulation by placing a given number of        //
// electrons and ions at random positions between the electrodes        //
//----------------------------------------------------------------------//

func (sim *SimulationState) InitParticles(nseed int) {
	for i := 0; i < nseed; i++ {
		sim.X_e[i] = L * sim.R01() // initial random position of the electron
		sim.Vx_e[i] = 0
		sim.Vy_e[i] = 0
		sim.Vz_e[i] = 0            // initial velocity components of the electron
		sim.X_i[i] = L * sim.R01() // initial random position of the ion
		sim.Vx_i[i] = 0
		sim.Vy_i[i] = 0
		sim.Vz_i[i] = 0 // initial velocity components of the ion
	}
	sim.N_e = nseed // initial number of electrons
	sim.N_i = nseed // initial number of ions
}

//----------------------------------------------------------------------//
// Null-Collision precomputation and sampling methods                   //
//----------------------------------------------------------------------//

func (sim *SimulationState) InitNullCollision() {
	sim.NuStarE = sim.MaxElectronCollFreq()
	sim.PStarE = 1.0 - math.Exp(-sim.NuStarE*DT_E)

	sim.NuStarI = sim.MaxIonCollFreq()
	sim.PStarI = 1.0 - math.Exp(-sim.NuStarI*DT_I)

	fmt.Printf(">> GoPIC: null-collision: nu*_e = %e, P*_e = %e\n", sim.NuStarE, sim.PStarE)
	fmt.Printf(">> GoPIC: null-collision: nu*_i = %e, P*_i = %e\n", sim.NuStarI, sim.PStarI)
}

func (sim *SimulationState) sampleBinomial(n int, p float64) int {
	if n <= 0 || p <= 0.0 {
		return 0
	}
	if p >= 1.0 {
		return n
	}
	mean := float64(n) * p
	variance := mean * (1.0 - p)
	if mean >= 5.0 && variance > 0.0 {
		// de Moivre-Laplace Gaussian approximation
		val := int(math.Round(mean + sim.Rng.NormFloat64()*math.Sqrt(variance)))
		if val < 0 {
			return 0
		}
		if val > n {
			return n
		}
		return val
	}

	count := 0
	for i := 0; i < n; i++ {
		if sim.R01() < p {
			count++
		}
	}
	return count
}

// randomSample selects k unique indices from [0, n) using partial Fisher-Yates shuffle.
// NOTE: In this step (2.algorithmic-port), memory is allocated via make([]int, n)
// to faithfully represent the direct algorithmic port from C++ std::vector<int>
// before introducing zero-allocation buffer reuse in Step 3.
func (sim *SimulationState) randomSample(n, k int) []int {
	if k <= 0 || n <= 0 {
		return nil
	}
	if k > n {
		k = n
	}
	pool := make([]int, n)
	for i := 0; i < n; i++ {
		pool[i] = i
	}
	for i := 0; i < k; i++ {
		j := i + int(sim.R01()*float64(n-i))
		pool[i], pool[j] = pool[j], pool[i]
	}
	return pool[:k]
}

//---------------------------------------------------------------------//
// simulation of one radiofrequency cycle                              //
//---------------------------------------------------------------------//

func (sim *SimulationState) Step1ComputeElectronDensity() {
	for p := 0; p < N_G; p++ {
		sim.E_density[p] = 0 // electron density - computed in every time step
	}
	for k := 0; k < sim.N_e; k++ {
		c0 := sim.X_e[k] * INV_DX
		p := int(c0)
		c2 := c0 - float64(p)
		w2 := c2 * FACTOR_W
		w1 := FACTOR_W - w2
		sim.E_density[p] += w1
		sim.E_density[p+1] += w2
	}
	sim.E_density[0] *= 2.0
	sim.E_density[N_G-1] *= 2.0
	for p := 0; p < N_G; p++ {
		sim.Cumul_e_density[p] += sim.E_density[p]
	}
}

func (sim *SimulationState) Step1ComputeIonDensity(t int) {
	if (t % N_SUB) == 0 { // ion density - computed in every N_SUB-th time steps (subcycling)
		for p := 0; p < N_G; p++ {
			sim.I_density[p] = 0
		}
		for k := 0; k < sim.N_i; k++ {
			c0 := sim.X_i[k] * INV_DX
			p := int(c0)
			c2 := c0 - float64(p)
			w2 := c2 * FACTOR_W
			w1 := FACTOR_W - w2
			sim.I_density[p] += w1
			sim.I_density[p+1] += w2
		}
		sim.I_density[0] *= 2.0
		sim.I_density[N_G-1] *= 2.0
	}
	for p := 0; p < N_G; p++ {
		sim.Cumul_i_density[p] += sim.I_density[p]
	}
}

func (sim *SimulationState) Step2SolvePoisson(currentTime float64) {
	var rho Xvector
	for p := 0; p < N_G; p++ {
		rho[p] = E_CHARGE * (sim.I_density[p] - sim.E_density[p]) // get charge density
	}
	sim.SolvePoisson(&rho, currentTime) // compute potential and electric field
}

func (sim *SimulationState) Step3MoveElectrons(t_index int) {
	if !sim.Measurement_mode {
		// Fast-path for non-diagnostic cycles (1-mul CIC FMA)
		for k := 0; k < sim.N_e; k++ {
			c0 := sim.X_e[k] * INV_DX
			p := int(c0)
			c2 := c0 - float64(p)
			ep := sim.Efield[p]
			e_x := ep + c2*(sim.Efield[p+1]-ep)

			sim.Vx_e[k] -= e_x * FACTOR_E
			sim.X_e[k] += sim.Vx_e[k] * DT_E
		}
		return
	}

	for k := 0; k < sim.N_e; k++ { // move all electrons in every time step
		c0 := sim.X_e[k] * INV_DX
		p := int(c0)
		c2 := c0 - float64(p)
		c1 := 1.0 - c2
		ep := sim.Efield[p]
		e_x := ep + c2*(sim.Efield[p+1]-ep)

		// measurements: 'x' and 'v' are needed at the same time, i.e. old 'x' and mean 'v'
		mean_v := sim.Vx_e[k] - 0.5*e_x*FACTOR_E
		sim.Counter_e_xt[p][t_index] += c1
		sim.Counter_e_xt[p+1][t_index] += c2
		sim.Ue_xt[p][t_index] += c1 * mean_v
		sim.Ue_xt[p+1][t_index] += c2 * mean_v
		v_sqr := mean_v*mean_v + sim.Vy_e[k]*sim.Vy_e[k] + sim.Vz_e[k]*sim.Vz_e[k]
		energy := (0.5 * E_MASS * INV_EV_TO_J) * v_sqr
		sim.Meanee_xt[p][t_index] += c1 * energy
		sim.Meanee_xt[p+1][t_index] += c2 * energy
		energy_index := minInt(int(v_sqr*FACTOR_ENERGY_E+0.5), CS_RANGES-1)
		velocity := math.Sqrt(v_sqr)
		rate := sim.Sigma[E_ION][energy_index] * velocity * DT_E * GAS_DENSITY
		sim.Ioniz_rate_xt[p][t_index] += c1 * rate
		sim.Ioniz_rate_xt[p+1][t_index] += c2 * rate

		// measure EEPF in the center
		if (MIN_X < sim.X_e[k]) && (sim.X_e[k] < MAX_X) {
			eepf_index := int(energy * INV_DE_EEPF)
			if eepf_index < N_EEPF {
				sim.Eepf[eepf_index] += 1.0
			}
			sim.Mean_energy_accu_center += energy
			sim.Mean_energy_counter_center++
		}

		// update velocity and position
		sim.Vx_e[k] -= e_x * FACTOR_E
		sim.X_e[k] += sim.Vx_e[k] * DT_E
	}
}

func (sim *SimulationState) Step4MoveIons(t_index, t int) {
	if (t % N_SUB) != 0 {
		return
	}

	if !sim.Measurement_mode {
		// Fast-path for non-diagnostic cycles (1-mul CIC FMA)
		for k := 0; k < sim.N_i; k++ {
			c0 := sim.X_i[k] * INV_DX
			p := int(c0)
			c2 := c0 - float64(p)
			ep := sim.Efield[p]
			e_x := ep + c2*(sim.Efield[p+1]-ep)

			sim.Vx_i[k] += e_x * FACTOR_I
			sim.X_i[k] += sim.Vx_i[k] * DT_I
		}
		return
	}

	for k := 0; k < sim.N_i; k++ {
		c0 := sim.X_i[k] * INV_DX
		p := int(c0)
		c2 := c0 - float64(p)
		c1 := 1.0 - c2
		ep := sim.Efield[p]
		e_x := ep + c2*(sim.Efield[p+1]-ep)

		// measurements: 'x' and 'v' are needed at the same time, i.e. old 'x' and mean 'v'
		mean_v := sim.Vx_i[k] + 0.5*e_x*FACTOR_I
		sim.Counter_i_xt[p][t_index] += c1
		sim.Counter_i_xt[p+1][t_index] += c2
		sim.Ui_xt[p][t_index] += c1 * mean_v
		sim.Ui_xt[p+1][t_index] += c2 * mean_v
		v_sqr := mean_v*mean_v + sim.Vy_i[k]*sim.Vy_i[k] + sim.Vz_i[k]*sim.Vz_i[k]
		energy := (0.5 * AR_MASS * INV_EV_TO_J) * v_sqr
		sim.Meanei_xt[p][t_index] += c1 * energy
		sim.Meanei_xt[p+1][t_index] += c2 * energy

		// update velocity and position
		sim.Vx_i[k] += e_x * FACTOR_I
		sim.X_i[k] += sim.Vx_i[k] * DT_I
	}
}

func (sim *SimulationState) Step5CheckBoundariesElectrons() {
	var k int = 0
	var out bool
	for k < sim.N_e { // check boundaries for all electrons in every time step
		out = false
		if sim.X_e[k] < 0 {
			sim.N_e_abs_pow++ // the electron is out at the powered electrode
			out = true
		}
		if sim.X_e[k] > L {
			sim.N_e_abs_gnd++ // the electron is out at the grounded electrode
			out = true
		}
		if out { // remove the electron, if out
			sim.X_e[k] = sim.X_e[sim.N_e-1]
			sim.Vx_e[k] = sim.Vx_e[sim.N_e-1]
			sim.Vy_e[k] = sim.Vy_e[sim.N_e-1]
			sim.Vz_e[k] = sim.Vz_e[sim.N_e-1]
			sim.N_e--
		} else {
			k++
		}
	}
}

func (sim *SimulationState) Step6CheckBoundariesIons(t int) {
	if (t % N_SUB) != 0 {
		return
	}

	var k, energy_index int
	var out bool
	var v_sqr float64

	k = 0
	for k < sim.N_i {
		out = false
		if sim.X_i[k] < 0 { // the ion is out at the powered electrode
			sim.N_i_abs_pow++
			out = true
			v_sqr = sim.Vx_i[k]*sim.Vx_i[k] + sim.Vy_i[k]*sim.Vy_i[k] + sim.Vz_i[k]*sim.Vz_i[k]
			energy_index = int(v_sqr * FACTOR_ENERGY_IFED)
			if energy_index < N_IFED {
				sim.Ifed_pow[energy_index]++ // save IFED at the powered electrode
			}
		}
		if sim.X_i[k] > L { // the ion is out at the grounded electrode
			sim.N_i_abs_gnd++
			out = true
			v_sqr = sim.Vx_i[k]*sim.Vx_i[k] + sim.Vy_i[k]*sim.Vy_i[k] + sim.Vz_i[k]*sim.Vz_i[k]
			energy_index = int(v_sqr * FACTOR_ENERGY_IFED)
			if energy_index < N_IFED {
				sim.Ifed_gnd[energy_index]++ // save IFED at the grounded electrode
			}
		}
		if out { // delete the ion, if out
			sim.X_i[k] = sim.X_i[sim.N_i-1]
			sim.Vx_i[k] = sim.Vx_i[sim.N_i-1]
			sim.Vy_i[k] = sim.Vy_i[sim.N_i-1]
			sim.Vz_i[k] = sim.Vz_i[sim.N_i-1]
			sim.N_i--
		} else {
			k++
		}
	}
}

func (sim *SimulationState) Step7CollisionsElectrons() {
	if sim.N_e <= 0 {
		return
	}

	numColl := sim.sampleBinomial(sim.N_e, sim.PStarE)
	if numColl <= 0 {
		return
	}

	candidates := sim.randomSample(sim.N_e, numColl)
	for _, ki := range candidates {
		v_sqr := sim.Vx_e[ki]*sim.Vx_e[ki] + sim.Vy_e[ki]*sim.Vy_e[ki] + sim.Vz_e[ki]*sim.Vz_e[ki]
		velocity := math.Sqrt(v_sqr)
		energy_index := minInt(int(v_sqr*FACTOR_ENERGY_E+0.5), CS_RANGES-1)

		realNu := sim.SigmaTotE[energy_index] * velocity
		pAccept := realNu / sim.NuStarE
		if pAccept > 1.0 {
			pAccept = 1.0
		}

		if sim.R01() < pAccept {
			sim.CollisionElectron(sim.X_e[ki], &sim.Vx_e[ki], &sim.Vy_e[ki], &sim.Vz_e[ki], energy_index)
			sim.N_e_coll++
		}
	}
}

func (sim *SimulationState) Step8CollisionIons(t int) {
	if (t%N_SUB) != 0 || sim.N_i <= 0 {
		return
	}

	numColl := sim.sampleBinomial(sim.N_i, sim.PStarI)
	if numColl <= 0 {
		return
	}

	candidates := sim.randomSample(sim.N_i, numColl)
	for _, ki := range candidates {
		vx_a := sim.RMB()
		vy_a := sim.RMB()
		vz_a := sim.RMB()
		gx := sim.Vx_i[ki] - vx_a
		gy := sim.Vy_i[ki] - vy_a
		gz := sim.Vz_i[ki] - vz_a
		g_sqr := gx*gx + gy*gy + gz*gz
		g := math.Sqrt(g_sqr)
		energy_index := minInt(int(g_sqr*FACTOR_ENERGY_I+0.5), CS_RANGES-1)

		realNu := sim.SigmaTotI[energy_index] * g
		pAccept := realNu / sim.NuStarI
		if pAccept > 1.0 {
			pAccept = 1.0
		}

		if sim.R01() < pAccept {
			sim.CollisionIon(&sim.Vx_i[ki], &sim.Vy_i[ki], &sim.Vz_i[ki], &vx_a, &vy_a, &vz_a, energy_index)
			sim.N_i_coll++
		}
	}
}

func (sim *SimulationState) Step9CollectXtData(t_index int) {
	if !sim.Measurement_mode {
		return
	}

	for p := 0; p < N_G; p++ {
		sim.Pot_xt[p][t_index] += sim.Pot[p]
		sim.Efield_xt[p][t_index] += sim.Efield[p]
		sim.Ne_xt[p][t_index] += sim.E_density[p]
		sim.Ni_xt[p][t_index] += sim.I_density[p]
	}
}

func (sim *SimulationState) DoOneCycle() {
	var t int
	var t_index int

	for t = 0; t < N_T; t++ { // the RF period is divided into N_T equal time intervals (time step DT_E)
		sim.Time += DT_E    // update of the total simulated time
		t_index = t / N_BIN // index for XT distributions

		sim.Step1ComputeElectronDensity()
		sim.Step1ComputeIonDensity(t)
		sim.Step2SolvePoisson(sim.Time)

		sim.Step3MoveElectrons(t_index)
		sim.Step4MoveIons(t_index, t)

		sim.Step5CheckBoundariesElectrons()
		sim.Step6CheckBoundariesIons(t)

		sim.Step7CollisionsElectrons()
		sim.Step8CollisionIons(t)

		sim.Step9CollectXtData(t_index)

		if (t % 1000) == 0 {
			fmt.Printf(" c = %8d  t = %8d  #e = %8d  #i = %8d\n", sim.Cycle, t, sim.N_e, sim.N_i)
		}
	}
	fmt.Fprintf(sim.Datafile, "%8d  %8d  %8d\n", sim.Cycle, sim.N_e, sim.N_i)
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}
